DOCS-2343 Spark History Server Updates (#79)
lucidworks/fusion-cloud-native: survival_guide/3_operations.adoc

==== Spark History Server

//tag::history-intro[]

> While logs from the Spark driver and executor pods can be viewed using `kubectl logs [POD_NAME]`, executor pods are deleted at their end of their execution, and driver pods are deleted by Fusion on a default schedule of every hour. In order to store and view Spark logs in a more long-term fashion, you can install the https://spark.apache.org/docs/latest/monitoring.html[Spark History Server] into your Kubernetes cluster and configure Spark to write logs in a manner that will persist.

Reviewer: I suggest using a caret (^) to make external links open in a new window/tab, as in `[Spark History Server^]`.

[source,yaml]
----
gcs:
  enableGCS: true
  secret: history-secrets
  key: sparkhistory.json
  logDirectory: gs://[BUCKET_NAME]
service:
  type: ClusterIP
  port: 18080
----

> The `service` key sets up the history server on an internal IP within your cluster but does not create an external LoadBalancer (which is the default setting in the Spark History Server Helm chart). This prevents the server from being exposed to outside access by default. We’ll look at how to access the history server shortly.

Reviewer 2: Don't need "by default", it confuses our recommended default (which isn't really a default, it's just our recommendation) with the spark history server default in their helm chart. Also reduced the needed context for each sentence. And we mark this as a "note" because it's not essential to understand this in order to follow the steps.
```suggestion
Note that, by default, the Spark History Server Helm chart creates an external LoadBalancer, exposing it to outside access. This is usually undesirable. In the above, we prevent this via the `service` key - the history server will only be set up on an internal IP within your cluster and will not be exposed externally. Later, we will show how to properly access the history server.
```

> The `key` and `secret` fields provide the Spark History Server with the details of where it will find an account with access to the Google Cloud Storage bucket given in `logDirectory`. In the following example, we're going to set up a new service account that will be shared between the Spark History Server and the Spark driver/executors for both viewing and writing logs.

Reviewer 2: Be consistent with what we are calling it on this page after it's introduced - i.e. "Spark History Server" or "history server". use the same term every time with the same capitalization. This applies to the whole page not just this usage.

[source,bash]
----
$ export ACCOUNT_NAME=sparkhistory
$ export GCP_PROJECT_ID=[PROJECT_ID]
$ gcloud iam service-accounts create ${ACCOUNT_NAME} --display-name "${ACCOUNT_NAME}"
$ gcloud iam service-accounts keys create "${ACCOUNT_NAME}.json" --iam-account "${ACCOUNT_NAME}@${GCP_PROJECT_ID}.iam.gserviceaccount.com"
$ gcloud projects add-iam-policy-binding ${GCP_PROJECT_ID} --member "serviceAccount:${ACCOUNT_NAME}@${GCP_PROJECT_ID}.iam.gserviceaccount.com" --role roles/storage.admin
$ gsutil iam ch serviceAccount:${ACCOUNT_NAME}@${GCP_PROJECT_ID}.iam.gserviceaccount.com:objectAdmin gs://[BUCKET_NAME]
----

> First, we use `create` to create a new service account, and then `keys create` to create a JSON keypair which we will shortly upload into our cluster as a Kubernetes secret. We then give our service account the `storage/admin` role, allowing it to perform create and view operations, and the final `gsutil` command applies our service account to our chosen bucket. If you have an existing service account you wish to use instead, you can skip the `create` command, though you will still need to create the JSON keypair and ensure that the existing account can read and write to the log bucket.

Reviewer 2: When you give command line things to run, the explanation for what it's going to do should be before each command, not after. Basically "here's what we're gonna do next" and then "here's us doing it", rather than "here's us doing something you don't understand" and then "here's what we just did". 

It can be broken up into smaller paragraphs (explanation -> command, explanation -> command) to help with readability also.

$ gcloud projects add-iam-policy-binding ${GCP_PROJECT_ID} --member "serviceAccount:${ACCOUNT_NAME}@${GCP_PROJECT_ID}.iam.gserviceaccount.com" --role roles/storage.admin
$ gsutil iam ch serviceAccount:${ACCOUNT_NAME}@${GCP_PROJECT_ID}.iam.gserviceaccount.com:objectAdmin gs://[BUCKET_NAME]
----

First, we use `create` to create a new service account, and then `keys create` to create a JSON keypair which we will shortly upload into our cluster as a Kubernetes secret. We then give our service account the `storage/admin` role, allowing it to perform create and view operations, and the final `gsutil` command applies our service account to our chosen bucket. If you have an existing service account you wish to use instead, you can skip the `create` command, though you will still need to create the JSON keypair and ensure that the existing account can read and write to the log bucket.

We now need to upload the JSON keypair into the cluster as a secret:

[source,bash]
----
kubectl -n [NAMESPACE] create secret generic history-secrets --from-file=sparkhistory.json
----

> With all this in place, the Spark History Server can now be installed with `helm install stable/spark-history-server --values values.yaml`.

Reviewer 2: This command will fail, asking you to specify the name. We should suggest (namespace)-spark-history-server.
`helm install [NAMESPACE]-spark-history-server stable/spark-history-server --values values.yaml`

[source,yaml]
----
gcs:
  enableGCS: true
  secret: history-secrets
  key: sparkhistory.json
  logDirectory: gs://[BUCKET_NAME]
service:
  type: ClusterIP
  port: 18080

> nfs.enableExampleNFS: false

Reviewer 2: for some reason this doesn't actually disable the NFS pod! You need to use a different notation.
```suggestion
nfs:
  enableExampleNFS: false
```

Reviewer 2: Sorry, you also need to add another thing or pod won't schedule
```suggestion
pvc:
  enablePVC: false
nfs:
  enableExampleNFS: false
```

[source,yaml]
----
s3:
  enableS3: true
  enableIAM: false
  accessKeyName: aws-access-key
  secretKeyName: aws-secret-key
  logDirectory: s3a://[BUCKET_NAME]
----

===== Configuring Spark

> After the History Server has been set up, then the config map for Fusion's job-launcher will need its `application.yaml` key will need to be updated with these Spark settings so the driver and executors know where to write out their logs.

Reviewer 2: ```suggestion
After starting the history server, we must update the config map for Fusion's job-launcher so it can write logs to the same location that Spark History Server is reading from.
```

> In this example, having installed Fusion into a namespace of `sparkhistory`, we will edit the config map to write the logs to a Google Cloud Storage bucket.

Reviewer 2: ```suggestion
In this example, having installed Fusion into a namespace of `sparkhistory`, we will edit the config map to write the logs to the same Google Cloud Storage bucket we configured the history server to read from.
```

[source,yaml]
----
s3:
  enableS3: true
  enableIAM: false
  accessKeyName: aws-access-key
  secretKeyName: aws-secret-key
  logDirectory: s3a://[BUCKET_NAME]
----

===== Configuring Spark

After the History Server has been set up, then the config map for Fusion's job-launcher will need its `application.yaml` key will need to be updated with these Spark settings so the driver and executors know where to write out their logs.

In this example, having installed Fusion into a namespace of `sparkhistory`, we will edit the config map to write the logs to a Google Cloud Storage bucket.

[source,bash]
----

> kubectl edit cm spark-history-job-launcher

Reviewer 2: shouldn't it be `sparkhistory-job-launcher`

----

> Update the `spark` key with these settings and then restart the `job-launcher` pod.

Reviewer 2: might want to be more specific about restarting the pod - I think you just mean deleting it

[source,yaml]
----
spark:
  …
  eventLog:
    enabled: true
    compress: true
    dir: gs://[BUCKET_NAME]
  hadoop:
    google:
      cloud:
        auth:
          service:
            account:
              json:
                keyfile: /etc/secrets/[ACCOUNT_NAME].json
  kubernetes:
    driver:
      secrets:
        history-secrets: /etc/secrets
    executor:
      secrets:
        history-secrets: /etc/secrets
  …
----

> 

Reviewer 2: We need to add:
1. How to go verify that it's working - run a job and check the history server.
2. How to actually reach and expose the spark history server.

[source,yaml]
----
s3:
  enableS3: true
  enableIAM: false
  accessKeyName: aws-access-key
  secretKeyName: aws-secret-key
  logDirectory: s3a://[BUCKET_NAME]
----

===== Configuring Spark

After starting the Spark History Server, we must update the config map for Fusion's job-launcher so it can write logs to the same location that Spark History Server is reading from.

In this example, having installed Fusion into a namespace of `sparkhistory`, we will edit the config map to write the logs to the same Google Cloud Storage bucket we configured the Spark History Server to read from.

[source,bash]
----

> kubectl edit cm sparkhistory-job-launcher

Reviewer 2: I did something apparently wrong when editing configmap and now this command is printing out newline characters and is a PITA to edit. I might suggest we tell people to make a backup of the configmap before editing or tell them how to restore it if something goes wrong.

----

Update the `spark` key with these settings and then restart the `job-launcher` pod.

[source,yaml]
----
spark:
  …
  eventLog:
    enabled: true
    compress: true
    dir: gs://[BUCKET_NAME]
  hadoop:
    google:
      cloud:
        auth:
          service:
            account:
              json:
                keyfile: /etc/secrets/[ACCOUNT_NAME].json
  kubernetes:
    driver:
      secrets:

>         history-secrets: /etc/secrets

Reviewer 2: pretty sure using /etc/secrets conflicts with some other secret (service account key?)
I think you need to use /etc/history-secrets or something unique instead for all of these instead of /etc/secrets
